onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -216,6 +265,128 @@
+  API_IMPL_END
+}
+
+ORT_API_STATUS_IMPL(OrtSessionOptionsAppendExecutionProvider_DML2, _In_ OrtSessionOptions* options, OrtDmlDeviceOptions* device_opts) {
+API_IMPL_BEGIN
+  //options->provider_factories.push_back(onnxruntime::DMLProviderFactoryCreator::Create(device_id));
+  
+    OrtDmlPerformancePreference p = device_opts->p;
+    OrtDmlDeviceFilter f = device_opts->f;
+
+    // Create DXCore Adapter Factory
+    ComPtr<IDXCoreAdapterFactory> adapterFactory;
+    ORT_THROW_IF_FAILED(::DXCoreCreateAdapterFactory(adapterFactory.GetAddressOf()));
+
+    // Get a list of all the adapters that support compute
+    ComPtr<IDXCoreAdapterList> d3D12CoreComputeAdapters;
+    GUID attributes[]{ DXCORE_ADAPTER_ATTRIBUTE_D3D12_CORE_COMPUTE };
+    ORT_THROW_IF_FAILED(
+        adapterFactory->CreateAdapterList(_countof(attributes),
+            attributes,
+            d3D12CoreComputeAdapters.GetAddressOf()));
+
+    const uint32_t count{ d3D12CoreComputeAdapters->GetAdapterCount() };
+    int compute_only_device_id = -1;
+
+    // will hold all the adapters in the order specified by the device options
+    ComPtr<IDXCoreAdapter>* ordered_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int ordered_adapter_count = 0;
+
+    // Used in the case that both GPU and NPU adapters are considered
+    ComPtr<IDXCoreAdapter>* gpu_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int gpu_adapter_count = 0;
+    ComPtr<IDXCoreAdapter>* npu_adapters =
+        (ComPtr<IDXCoreAdapter>*) malloc(count * sizeof(ComPtr<IDXCoreAdapter>));
+    int npu_adapter_count = 0;
+
+    // Iterate through all compute capable adapters
+    for (uint32_t i = 0; i < count; ++i)
+    {
+        ComPtr<IDXCoreAdapter> candidateAdapter;
+        ORT_THROW_IF_FAILED(
+            d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.GetAddressOf()));
+
+        // Only considering hardware adapters
+        if (!isHardwareAdapter(candidateAdapter))
+            continue;
+
+        if (f == OrtDmlDeviceFilter::Gpu) // consider GPUs only
+        {
+            if (supportsGraphics(candidateAdapter)) {
+                ordered_adapters[ordered_adapter_count] = candidateAdapter;
+                ordered_adapter_count++;
+            }
+        }
+        else if(f == OrtDmlDeviceFilter::Npu) // consider NPUs only
+        {
+            if (!supportsGraphics(candidateAdapter)) {
+                ordered_adapters[ordered_adapter_count] = candidateAdapter;
+                ordered_adapter_count++;
+            }
+        }
+        else // consider both GPUs and NPUs
+        {
+            if (supportsGraphics(candidateAdapter)) {
+                gpu_adapters[gpu_adapter_count] = candidateAdapter;
+                gpu_adapter_count++;
+            } else {
+                npu_adapters[npu_adapter_count] = candidateAdapter;
+                npu_adapter_count++;
+            }
+        }
```

Comment:
May be better to do something like this:

enum class DeviceType { GPU, NPU, BadDevice};
struct AdapterInfo {
  ComPtr<IDXCoreAdapter> Adapter;
  DeviceType Type; // GPU or NPU
}

......


auto selected_adapters = std::vector<AdapterInfo>();
std::function<DeviceType (IDXCoreAdapter*)> filterQueryCheck[4] = 
{
  nullptr, // enum cannot take 0 value
  [](IDXCoreAdapter* adapter){ return IsGPU(adapter) ? DeviceType::GPU : DeviceType::BadDevice; },
  [](IDXCoreAdapter* adapter){ return IsNPU(adapter) ? DeviceType::NPU : DeviceType::BadDevice; },
  [](IDXCoreAdapter* adapter){ return IsGPU(adapter) ? DeviceType::GPU :
                                                                                               IsNPU(adapter) ? DeviceType::NPU : DeviceType::BadDevice; }
}

...
for (uint32_t i = 0; i < count; ++i)
{
    ComPtr<IDXCoreAdapter> candidateAdapter;
        ORT_THROW_IF_FAILED(
            d3D12CoreComputeAdapters->GetAdapter(i, candidateAdapter.GetAddressOf()));

    auto adapterType = filterQueryCheck[filter](candidateAdapter.Get());
    if (adapterType != DeviceType::BadDevice) {
        selected_adapters.push_back({candidateAdapter, adapterType })
    }
}